Prevent font upload if max7456 is not detected

etracer65 opened the pull request:
Fixes #9155 

Prevents wedge if a font upload is attempted but the max7456 device wasn't detected during initialization.

As part of #9013 logic was added to detect whether or not the max7456 was actually detected. If not then configuration of the SPI bus was not completed. Unfortunately the font upload process didn't consider whether the max7456 was available and always tried to communicate with the device. Since the SPI bus configuration wasn't completed this led to a wedge.

This problem could be replicated by using any flight controller that doesn't properly power the max7456 when on USB only. An example is the DALRCF722DUAL. Now for devices like this the font upload will be prevented and the font manager button in the Configurator will be hidden.

Tested with a variety of boards that properly power the max7456 on USB such as OMNIBUSF4SD, SPRACINGF7DUAL, MATEKF722SE.

Tested on a DALRCF722DUAL board that **does not** power the max7456 from USB. When only connected to USB the "Font Manager" functionality is hidden. When connected to battery power then the device is detected and fonts can be uploaded normally.

mikeller commented on src/main/msp/msp.c:
```diff
@@ -802,7 +802,9 @@ static bool mspCommonProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst, mspPostProce
 #ifdef USE_MAX7456
-        osdFlags |= OSD_FLAGS_OSD_HARDWARE_MAX_7456;
+        if (max7456IsDeviceDetected()) {
+            osdFlags |= OSD_FLAGS_OSD_HARDWARE_MAX_7456;
```
This should use a new flag - `OSD_FLAGS_OSD_HARDWARE_MAX_7456` is used to drive hiding / showing of all settings dependent on having a MAX7456 (powered or not) in configurator:

https://github.com/betaflight/betaflight-configurator/blob/53fb5944d14b9d66fd7fc7a47424c9ae075a26f4/src/js/tabs/osd.js#L1800

But the light of Unified Targets, we should probably make `OSD_FLAGS_OSD_HARDWARE_MAX_7456` dynamic, and dependent on `max7456Config->csTag && max7456Config->spiDevice`, to allow proper detection if a MAX7456 device is actually configured.

etracer65 replied:
I'm not sure I fully understand. In the Configurator the only logic that uses this flag (`haveMax7456Video`) is that which hides the font manager division. We don't seem to have any other logic based on actual detection of the max7456. Also `max7456IsDeviceDetected()` only returns try if the max7456 successfully completes detection and initialization. I don't see why checking for `max7456Config->csTag && max7456Config->spiDevice` is better.

mikeller replied:
> In the Configurator the only logic that uses this flag (haveMax7456Video) is that which hides the font manager division.

That's incorrect - there are other parts of the configuration screen that are dependent on `requires-max7456`:
- video mode selection;
- active OSD profile selection (this is probably a bug though).

Requiring that a battery is plugged in on some flight controllers just to be able to configure the video format is contrary to doing this in the safest possible way.
Also, having a way to figure out in configurator that a MAX7456 is configured, but not powered will be helpful to display an appropriate notice to users re how they can flash fonts.

In general, there are 3 different possible configurations for MAX7456, and I think having two flags to encode them all makes sense:
- no MAX7456 configured;
- MAX7456 configured but not powered up;
- MAX7456 configured and powered up.

mikeller commented on src/main/drivers/max7456.c:
```diff
@@ -484,15 +487,15 @@ bool max7456Init(const max7456Config_t *max7456Config, const vcdProfile_t *pVcdP
 
     uint8_t osdm = max7456Send(MAX7456ADD_OSDM|MAX7456ADD_READ, 0xff);
 
+    __spiBusTransactionEnd(busdev);
```
Is this a separate bugfix, and should this go into `4.1-maintenance`?

etracer65 replied:
No, it's only applicable to 4.2. The code introduced in #9013 would return without ending the transaction if the max7456 check failed. So if something else shared the bus with the max7456 it would be affected.

etracer65 commented:
Added a separate flag that determines whether the max7456 was actually detected (`OSD_FLAGS_MAX7456_DETECTED`). Also updated the existing `OSD_FLAGS_OSD_HARDWARE_MAX_7456` flag to be based on `max7456Config->csTag && max7456Config->spiDevice`.

Created a related Configurator PR to make use of the new `OSD_FLAGS_MAX7456_DETECTED` flag.

https://github.com/betaflight/betaflight-configurator/pull/1818
